specs/withdrawals.md

```diff
@@ -0,0 +1,181 @@
+
+[g-withdrawal]: glossary.md#withdrawal
+[g-mpt]: glossary.md#merkle-patricia-trie
+[g-relayer]: glossary.md#withdrawals
+[g-execution-engine]: glossary.md#execution-engine
+**Table of Contents**
+
+[Withdrawals][g-withdrawal] are cross domain transactions which are initiated on L2, and finalized by a transaction
+executed on L1. They may be used to transfer data and/or ETH from L1 to L2.
+
+**Vocabulary note**: *withdrawal* can refer to the transaction at various stages of the process, but we introduce
+more specific terms to differentiate between the transaction
+
+- *withdrawal initiating transaction* refers specifically to a transaction on L2 sent to the Withdrawals predeploy.
+- *withdrawal finalizing transaction* refers specifically to an L1 transaction which finalizes and relays the
+  withdrawal.
+
+Withdrawals are initiated on L2 via a call to the Withdrawals predeploy contract, which records the important properties
+of the message in its storage. Withdrawals are finalized on L1 via a call to the `L2WithdrawalVerifier` contract, which
+proves the inclusion of this withdrawal message.
+
+In this way, withdrawals are different from [deposits][g-deposits] which make use of a special transaction type in the
+[execution engine][g-execution-engine] client. Rather, withdrawals transaction must use smart contracts on L1 for
+finalization.
+
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+**Table of Contents**
+
+- [Withdrawal Flow](#withdrawal-flow)
+  - [On L2](#on-l2)
+  - [On L1](#on-l1)
+- [The L2 Withdrawer Contract](#the-l2-withdrawer-contract)
+- [The Optimism Portal Contract](#the-optimism-portal-contract)
+- [Withdrawal Verification and Finalization](#withdrawal-verification-and-finalization)
+- [Security Considerations](#security-considerations)
+  - [Key Properties of Withdrawal Verification](#key-properties-of-withdrawal-verification)
+  - [Handling Successfully Verified Messages That Fail When Relayed](#handling-successfully-verified-messages-that-fail-when-relayed)
+- [Summary of Definitions](#summary-of-definitions)
+  - [Constants](#constants)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Withdrawal Flow
+
+We first describe the end to end flow of initiating and finalizing a withdrawal:
+
+### On L2
+
+An L2 account sends a withdrawal message (and possibly also ETH) to the `Withdrawer` predeploy contract.
+   This is a very simple contract that stores the a hash of the withdrawal data.
+
+### On L1
+
+1. A [relayer][g-relayer] submits the required inputs (see below) to the `OptimismPortal` contract. The relayer need
+   not be the same entity which initiated the withdrawal on L2.
+2. The `OptimismPortal` contract retrieves the output root from the `OutputOracle`'s `l2Outputs()` function, and
+   performs the remainder of the verification process internally.
```

Comment:
This doesn't specify which output root is retrieved. Maybe specify in (1) that one of the input is a timestamp for which an L2 output root exists that commits to the withdrawal as registered on L2? Then say that we retrieve the output corresponding to the timestamp.

i.e. 

```
1. A [relayer][g-relayer] submits the required inputs to the `OptimismPortal` contract. The relayer need
   not be the same entity which initiated the withdrawal on L2.
   
   These inputs include the withdrawal transaction data, inclusion proofs, and a timestamp. The timestamp
   must be one for which an L2 output root exists, which commits to the withdrawal as registered on L2.
   More details are given below.

2. The `OptimismPortal` contract retrieves the output root for the given timestamp from the `OutputOracle`'s
   `l2Outputs()` function, and performs the remainder of the verification process internally.
 ```